// @oaas-hg/utils/modules/date-utils.js
module.exports = {
  dateDifferenceInDays,
  prettifyDate,
  diffTimestampAndToday,
  getDate,
  getCurrentMillis,
  dateToMillis,
};

const MILIS_PER_DAY = 86400000;
const MILIS_PER_HOUR = 3600000;
const GMT_TIME = Number(process.env.GMT_TIME) || -3;

function dateDifferenceInDays(date1, date2) {
  //Get each timestamp and substract them, the absolute value is the difference
  let difference = Math.abs(date2.getTime() - date1.getTime());
  //Divide the difference by the amount of miliseconds in a day
  difference = Math.round(difference / MILIS_PER_DAY);
  return difference;
}

function diffTimestampAndToday(timestamp) {
  //Today in theory should always be greater than the timestamp
  //Unless you have values from the future o.O

  return Math.abs(new Date().getTime() - timestamp);
}

function prettifyDate(date) {
  if (!date) date = new Date();
  //Adjust to GMT-3
  const fullDate = new Date(new Date(date).getTime() + MILIS_PER_HOUR * GMT_TIME);

  return (
    fullDate.toLocaleDateString("es-AR", {
      year: "2-digit",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
    }) + " hs"
  );
}

function getDate(timestamp = Date.now(), offset = -3) {
  //Adjust to GMT-3
  const fullDate = new Date(timestamp + MILIS_PER_HOUR * offset);

  return fullDate.toLocaleDateString("es-AR", {
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  });
}

function dateToMillis(year, month, day, hour, minute = 0, second = 0, gmt = "-03:00") {
  const stringDate = `${year}-${addZero(month)}-${addZero(day)}T${addZero(hour)}:${addZero(
    minute
  )}:${addZero(second)}${gmt}`;

  return new Date(stringDate).getTime() || undefined;
}

function addZero(num) {
  num = Number(num);
  return num >= 10 ? num : `0${num}`;
}

function getCurrentMillis() {
  return new Date().getTime() + MILIS_PER_HOUR * GMT_TIME;
}
